src/hooks/useSignChallenge.ts

```diff
@@ -0,0 +1,136 @@
+import { useState, useCallback } from 'react';
+import { useSignMessage } from 'wagmi';
+import { SIGNING_SERVICE_URL } from '../constants/constants';
+import { storageKeys } from '../constants/localStorage';
+import { SiweMessage } from 'siwe';
+import { DEFAULT_LOGIN_EXPIRATION_TIME_HOURS } from '../constants/constants';
+import { polygon } from 'wagmi/chains';
+
+export interface SiweSignatureData {
+  signatureSet: boolean;
+  expirationDate: string;
+}
+
+function createSiweMessage(address: `0x${string}`, nonce: string) {
+  // Make constants on config
+  const siweMessage = new SiweMessage({
+    scheme: 'https',
+    domain: window.location.host,
+    address,
+    statement: 'Please sign the message to login!',
+    uri: window.location.origin,
+    version: '1',
+    chainId: polygon.id,
+    nonce,
+    expirationTime: new Date(Date.now() + DEFAULT_LOGIN_EXPIRATION_TIME_HOURS * 60 * 60 * 1000).toISOString(), // Constructor in ms.
+  });
+  return siweMessage.toMessage();
+}
+
+export function useSiweSignature(address?: `0x${string}`) {
+  const { signMessageAsync } = useSignMessage();
+  const [signingPending, setSigningPending] = useState(false);
+
+  // Used to wait for the modal interaction and/or return of the
+  // signing promise.
+  const [signPromise, setSignPromise] = useState<{
+    resolve: () => void;
+    reject: (reason: Error) => void;
+  } | null>(null);
+
+  const storageKey = `${storageKeys.SIWE_SIGNATURE_KEY_PREFIX}${address}`;
+
+  const checkStoredSignature = useCallback((): SiweSignatureData | null => {
+    if (!address) return null;
+
+    try {
+      const stored = localStorage.getItem(storageKey);
+      if (!stored) return null;
+
+      const data: SiweSignatureData = JSON.parse(stored);
+      return new Date(data.expirationDate) > new Date() ? data : null;
+    } catch {
+      localStorage.removeItem(storageKey);
+      return null;
+    }
+  }, [address, storageKey]);
+
+  const signMessage = useCallback((): Promise<void> => {
+    return new Promise((resolve, reject) => {
```

Comment:
Should not be possible for it to be called at the same time with the current set-up, but you're right I will add it just in case to avoid potential future confusions.

Reply:
Yes, even it is not possible right now (which I agree), it is better to protect again in this call for the code to be more robust in case it is changed somewhere else.